bin/gen_tests.sh

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env sh
+
+# Purpose: Generate test scripts
+# Author : Álvaro Mondéjar Rubio
+# License: MIT
+# Date   : 2021 September 29th
+# Example:
+#
+#   $ sh ./bin/gen_tests.sh < lib/dpkg.txt > ./test.sh
+#   $ docker run --rm \
+#       -v $PWD/test.sh:/tmp/test.sh \
+#       -v $PWD/pacapt.dev:/usr/bin/pacman \
+#       -e "MSG_PREFIX=:: (INFO) " \
+#       ubuntu:14.04 \
+#       /bin/sh /tmp/test.sh 2>test.log
+
+echo "#!/bin/sh"
+echo ""
+echo "# Notes: This file is generated. Please don't modify them manually."
+echo ""
+echo "export PATH=/sbin:/bin:/usr/sbin:/usr/bin:\$PATH"
+echo ""
+echo "N_TEST=0  # Total number of tests"
+echo "N_FAIL=0  # Number of failed tests"
+echo "F_TMP=    # Temporary file"
+echo "T_FAIL=0  # Number of failed tests in current block"
+echo ""
+echo "set -u"
+echo ""
+echo ": \"\${GGREP:=grep}\""
+echo ""
+echo "if [ \"\$(uname)\" = \"SunOS\" ]; then"
+echo "  GGREP=ggrep"
+echo "fi"
+echo ""
+#
+# Just send message to STDERR
+#
+# When CI is set, we expect the current test script is executed directly
+# by Github-Runner shell (not via our Docker launcher.)
+# This is the case for non-docker tests (homebrew, pkgng).
+# We need to reduce the duplicate log message in these cases.
+#
+echo "_log()  { if [ -z \"\${CI:-}\" ]; then echo \"\$*\" 1>&2 ; fi; }"
+
+# A fancy wrappers of _log
+#
+# Where there is any message, we need to ensure they are on both
+# channels (STDIN, STDERR). Normally, the STDERR has more verbose
+# information, i.e, the output of all executions. The STDERR will
+# be [often] captured for later investigation and it's done via
+# the `tests/test.sh`. From STDIN we try to keep it compact.
+# ...
+#
+# shellcheck disable=SC2028
+echo "_fail() { _log \"\${MSG_PREFIX}Fail: \$*\"; printf \"\${MSG_PREFIX}\e[31mFail\e[0m: %s\n\" \"\$*\"; }" # red
```

Comment:
I think the better alternative of `puts` in Ruby is `echo -e`. Otherwise, the report when there is some failing test will be a bit hard to read. I've taken a screenshot [1].

Unfortunately `echo -e` is not POSIX compliant. And `printf "%s\n"` is not an immediate fix. Let me think if there is any trivial way.

[1] https://gist.github.com/icy/e6688bd112805831b7223dcdb1224c24#file-pacapt-pr-210-collapsed-png

Reply:
Not sure if I follow you. I'm not experiencing that using the terminal included by default on Ubuntu 20.04. 

<details>
<summary>apk.alpine:3.12.log</summary>

```
:: INFO ( apk vs. alpine:3.12 ) Exec: pacman -Qq
:: INFO ( apk vs. alpine:3.12 ) Fail: Expected ^bus2ybox
:: INFO ( apk vs. alpine:3.12 ) Info: Exec. output:
WARNING: Ignoring http://dl-cdn.alpinelinux.org/alpine/v3.12/main: No such file or directory
WARNING: Ignoring http://dl-cdn.alpinelinux.org/alpine/v3.12/community: No such file or directory
musl
busybox
alpine-baselayout
alpine-keys
libcrypto1.1
libssl1.1
ca-certificates-bundle
libtls-standalone
ssl_client
zlib
apk-tools
scanelf
musl-utils
libc-utils
:: INFO ( apk vs. alpine:3.12 ) Info: =====================================================
```

</details>

![image](https://user-images.githubusercontent.com/23049315/135590994-08a777df-7c13-4eaa-ad85-517bae745c1d.png)

Reply:
There is an actual problem with `| awk '{ if (NR <= 100) { printf(\" > %s\\\n\", \$0); }}"` and I think the fix can be easy now.

```
36,37c37,38
<       | awk '{ if (NR <= 100) { printf(" > %s\\n", $0); }}
<          END { if (NR > 100) { printf(" > ...\\n");}}'
---
>       | awk '{ if (NR <= 100) { printf(" > %s\n", $0); }}
>          END { if (NR > 100) { printf(" > ...\n");}}'
77c78
```

Reply:
So we expect `awk` to print `\n`, not `\\n`.

Reply:
but I don't really get why I got the issue, but you don't. I will look again after work.

Reply:
> There is an actual problem with `| awk '{ if (NR <= 100) { printf(\" > %s\\\n\", \$0); }}"` and I think the fix can be easy now.

I'm not able to reproduce it in my local environment:

```sh
$ sh bin/gen_tests.sh < tests/pkgng.txt > shell-pkgng.sh
$ ruby -n bin/gen_tests.rb < tests/pkgng.txt > ruby-pkgng.sh
$ diff ruby-pkgng.sh shell-pkgng.sh 
$ echo $?
0
```

I suppose that our environments differ in `echo` implementation.

Reply:
Can you replace `echo` with `/bin/echo`? that fixes the problem at my workstation (bash-5.1.8). `echo` is provided by your current shell.

Reply:
Wow, I didn't know that:

```sh
$ command -v echo
echo
```